Mathlib/RingTheory/Kaehler/JacobiZariski.lean

```diff
@@ -0,0 +1,567 @@
+      RingHom.id_apply, coe_eval₂Hom, IH, Hom.aeval_val, smul_add, map_aeval, tmul_add, tmul_smul,
+      ← @IsScalarTower.algebraMap_smul Q.Ring T, smul_zero, aeval_X, zero_add, Derivation.leibniz,
+      LinearEquiv.map_add, LinearEquiv.map_smul, Basis.repr_self, LinearMap.map_add, one_smul,
+      LinearMap.map_smul, Finsupp.linearCombination_single,
+      Function.comp_apply, ← cotangentSpaceBasis_apply]
+    rw [add_left_comm]
+    rfl
+
+lemma H1Cotangent.δAux_ofComp (x : (Q.comp P).Ring) :
+    δAux R Q ((Q.ofComp P).toAlgHom x) =
+      P.toKaehler.baseChange T (CotangentSpace.compCotangentSpace Q P
+        (1 ⊗ₜ[(Q.comp P).Ring] (D R (Q.comp P).Ring) x)).2 := by
+  letI : AddCommGroup (T ⊗[S] Ω[S⁄R]) := inferInstance
+  have : IsScalarTower (Q.comp P).Ring (Q.comp P).Ring T := IsScalarTower.left _
+  induction' x using MvPolynomial.induction_on with s x₁ x₂ hx₁ hx₂ p n IH
+  · simp only [algHom_C, δAux_C, sub_self, derivation_C, Derivation.map_algebraMap,
+      tmul_zero, map_zero, add_zero, MvPolynomial.algebraMap_apply, Prod.snd_zero]
+  · simp only [map_add, hx₁, hx₂, tmul_add, Prod.snd_add]
+  · simp only [map_mul, Hom.toAlgHom_X, ofComp_val, δAux_mul,
+      ← @IsScalarTower.algebraMap_smul Q.Ring T, algebraMap_apply, Hom.algebraMap_toAlgHom,
+      id.map_eq_id, map_aeval, RingHomCompTriple.comp_eq, comp_val, RingHom.id_apply, coe_eval₂Hom,
+      IH, Derivation.leibniz, tmul_add, tmul_smul, ← cotangentSpaceBasis_apply,
+      ← @IsScalarTower.algebraMap_smul (Q.comp P).Ring T, aeval_X, LinearEquiv.map_add,
+      LinearMapClass.map_smul, Prod.snd_add, Prod.smul_snd, LinearMap.map_add]
+    obtain (n | n) := n
+    · simp only [comp_vars, Sum.elim_inl, δAux_X, smul_zero, aeval_X,
+        CotangentSpace.compCotangentSpace, LinearEquiv.trans_apply, Basis.repr_symm_apply, zero_add,
+        Basis.repr_self, Finsupp.linearCombination_single, Basis.prod_apply, LinearMap.coe_inl,
+        LinearMap.coe_inr, Function.comp_apply, one_smul, map_zero]
+    · simp only [comp_vars, Sum.elim_inr, Function.comp_apply, algHom_C, δAux_C,
+        CotangentSpace.compCotangentSpace, LinearEquiv.trans_apply, Basis.repr_symm_apply,
+        algebraMap_smul, Basis.repr_self, Finsupp.linearCombination_single, Basis.prod_apply,
+        LinearMap.coe_inr, Basis.baseChange_apply, one_smul, LinearMap.baseChange_tmul,
+        toKaehler_cotangentSpaceBasis, add_left_inj, LinearMap.coe_inl]
+      rfl
+
+open Generators in
+/-- The connecting homomorphism in the Jacobi-Zariski sequence for a given family of generators. -/
+noncomputable
+def H1Cotangent.δ :
+    Q.H1Cotangent →ₗ[T] T ⊗[S] Ω[S⁄R] := by
+  classical
+  haveI : SMulCommClass (Q.comp P).Ring S T := by infer_instance
+  letI H := @SnakeLemma.δ' T _
+    (T ⊗[S] P.Cotangent) (Q.comp P).Cotangent Q.Cotangent
+    (T ⊗[S] P.CotangentSpace) (Q.comp P).CotangentSpace Q.CotangentSpace
+    _ _ _ _ _ _ _ _ _ _ _ _
+  haveI := H
```

Comment:
Is this needed because you can't provide a term directly?

Reply:
The last time I checked, Lean gets confused and it times out horribly, but it seems fine now.